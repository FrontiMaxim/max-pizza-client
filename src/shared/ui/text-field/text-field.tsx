import { forwardRef } from "react";
import cn from "classnames";

type TextFieldProps = {
  value?: string;
  placeholder?: string;
  type?: "text" | "password";
  name?: string;
  title: string;
  error?: string;
  className?: string;
  onChange?: () => void;
  onBlur?: () => void;
};

export const TextField = forwardRef<HTMLInputElement, TextFieldProps>(
  (props, ref) => {
    const {
      value,
      placeholder,
      title,
      type = "text",
      name,
      error,
      onChange,
      onBlur,
      className,
    } = props;

    return (
      <label className={cn("block", className)}>
        <div className="font-bold">{title}</div>
        <input
          value={value}
          type={type}
          name={name}
          placeholder={placeholder}
          onChange={onChange}
          onBlur={onBlur}
          ref={ref}
          className={cn(
            "mt-[5px] h-[48px] border rounded-[10px] pl-[19px] font-[16px] w-full",
            {
              ["border-error outline-none"]: error,
              ["border-grey outline-active"]: !error,
            }
          )}
        />
        {error && <div className="mt-[8px] text-error">{error}</div>}
      </label>
    );
  }
);
