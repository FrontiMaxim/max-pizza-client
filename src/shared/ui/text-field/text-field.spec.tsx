import { render, screen } from "@testing-library/react";
import { TextField } from "./text-field";
import { userEvent } from "@testing-library/user-event";

describe("TextField", () => {
  it("shound input is render", async () => {
    render(<TextField title="title" placeholder="placeholder" name="name" />);

    const inputElement = await screen.findByPlaceholderText("placeholder", {
      exact: true,
    });

    expect(inputElement).toBeInTheDocument();
    expect(inputElement).toBeVisible();
  });

  it("should value is 'text'", async () => {
    render(<TextField title="title" placeholder="placeholder" name="name" />);

    const inputElement = await screen.findByPlaceholderText("placeholder", {
      exact: true,
    });

    await userEvent.click(inputElement);
    await userEvent.keyboard("text");

    expect(inputElement).toHaveValue("text");
  });

  it("should focused", async () => {
    render(<TextField title="title" placeholder="placeholder" name="name" />);

    const inputElement = await screen.findByPlaceholderText("placeholder", {
      exact: true,
    });

    await userEvent.click(inputElement);

    expect(inputElement).toHaveFocus();
  });

  it("should error is render", async () => {
    render(
      <TextField
        title="title"
        placeholder="placeholder"
        name="name"
        error="error"
      />
    );

    const errorElement = await screen.findByText("error");

    expect(errorElement).toBeInTheDocument();
    expect(errorElement).toBeVisible();
  });

  it("should input`s view is default state", async () => {
    render(<TextField title="title" placeholder="placeholder" name="name" />);

    const inputElement = await screen.findByPlaceholderText("placeholder", {
      exact: true,
    });

    expect(inputElement).toMatchSnapshot();
  });

  it("should input`s view is error state", async () => {
    render(
      <TextField
        title="title"
        placeholder="placeholder"
        name="name"
        error="error"
      />
    );

    const inputElement = await screen.findByPlaceholderText("placeholder", {
      exact: true,
    });

    expect(inputElement).toMatchSnapshot();
  });
});
