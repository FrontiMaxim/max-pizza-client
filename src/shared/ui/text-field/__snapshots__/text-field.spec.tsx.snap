// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TextField > should input\`s view is default state 1`] = `
<input
  class="mt-[5px] h-[48px] border rounded-[10px] pl-[19px] font-[16px] w-full border-grey outline-active"
  name="name"
  placeholder="placeholder"
  type="text"
/>
`;

exports[`TextField > should input\`s view is error state 1`] = `
<input
  class="mt-[5px] h-[48px] border rounded-[10px] pl-[19px] font-[16px] w-full border-error outline-none"
  name="name"
  placeholder="placeholder"
  type="text"
/>
`;
